### PR TITLE
Add storage managment, upload & download

### PR DESCRIPTION
**Pull request by Gentux**
Add possibilities to user to upload or download files from their Nanocloud instances:
- Import Plaza  from community. 
- Add a storage container.
- Add Storage model for storage object.
- Create endpoints in StorageController:
  - POST api/upload : upload file in storage service,
  - GET api/files : get a files list,
  - GET api/files/token : get a 1 hour temporary token for download,
  - GET api/files/download : get a file from storage service.
- Add API test for new endpoints.
- Add Units test for Storage & Plaza services.

Replace PR #23

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -13,7 +13,7 @@ test-units:
 test-jshint:
 	jshint api/ config/
 
-test-licenses:
-	./tests/test-licenses.sh
+test-licences:
```
Comment (dynamiccast) [inline Makefile]: Why did you change the name ?

File: README.md [hunk 2]
```diff
@@ -26,6 +26,9 @@ Manual driver specific:
 - EXECUTION_SERVERS (mandatory) the IP of the execution server
 - WINDOWS_PASSWORD (mandatory) the Windows password for the *administrator* account
 
+Storage specific:
+- STORAGE_ADDRESS (mandatory, defaults to 'localhost') storage service's IP
```
Comment (dynamiccast) [inline README.md:30]: Is STORAGE_PORT still hard-coded ?

File: api/controllers/StorageController.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+ */
+
+/**
+ * UploadController
+ *
+ * @description :: Server-side logic for managing uploads
+ * @help        :: See http://sailsjs.org/#!/documentation/concepts/Controllers
+ */
+
+/* globals AccessToken */
```
Comment (dynamiccast) [inline api/controllers/StorageController.js]: You can separate globals name with comas.

File: api/controllers/StorageController.js [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+   */
+
+  upload: function(req, res) {
+    let user = req.user;
+
+    StorageService.findOrCreate(user)
+      .then((storage) => {
+        let filename = req.query["filename"];
+
+        req.file(filename).upload({
```
Comment (dynamiccast) [inline api/controllers/StorageController.js:51]: Use promise. If `req.file` does not return a promise, use bluebird's `promisify`

File: api/controllers/StorageController.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+          // If no files were uploaded, respond with an error.
+          if (uploadedFiles.length === 0){
+            return res.badRequest('No file was uploaded');
+          }
+
+          PlazaService.upload(
+              storage,
+              uploadedFiles[0],
+              (_, data) => {
+                res.send("Upload successful : " + data);
```
Comment (dynamiccast) [inline api/controllers/StorageController.js]: Use JsonApiService to respond with a valid JSON API response : `res.ok({})`

File: api/services/PlazaService.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+ * along with this program.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+const http = require("http");
+const fs = require("fs");
+
+module.exports = {
+
+  exec: function(hostname, cmd, stdin, callback) {
```
Comment (dynamiccast) [inline api/services/PlazaService.js]: All these function should be documented with YUI and return promises.

File: api/controllers/StorageController.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    StorageService.findOrCreate(user)
+      .then((storage) => {
+        PlazaService.files(storage, "", "/home/" + storage.username, (files) => {
+          res.send(files);
+        });
+      });
+  },
+
+  /**
+   * download a file fro; storage
```
Comment (dynamiccast) [inline api/controllers/StorageController.js]: Keep cql, qnd use azerty

File: api/controllers/StorageController.js [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+   *
+   * @method download
+   * @public true
+   */
+
+  download: function(req, res) {
+    let filename = req.query["filename"];
+    let downloadToken = req.query["token"];
+
+    AccessToken.findById(downloadToken.split(":")[0], (err, accessTokens) => {
```
Comment (dynamiccast) [inline api/controllers/StorageController.js]: Use promises

File: api/controllers/StorageController.js [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+    let timestamp = Date.now() / 1000;
+    let timeStone = timestamp + (3600 - timestamp % 3600);
+
+    AccessToken.findOne({userId: user.id}, (err, accessToken) => {
+      if (err !== null) {
+        res.negotiate(err);
+      }
+      res.send(200, {
+        token: accessToken.id + ":" + sha1(accessToken.token + ":" + filename + ":" + timeStone)
```
Comment (dynamiccast) [inline api/controllers/StorageController.js]: Should be in `StorageService` and properly unit tested.

File: api/policies/checkDownloadToken.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+  let timestamp = Date.now() / 1000;
+  let timeStone = timestamp + (3600 - timestamp % 3600);
+
+  AccessToken.findById(downloadToken.split(":")[0], (err, accessTokens) => {
+    if (err !== null) {
+      res.negotiate(err);
+    }
+    let accessToken = accessTokens[0];
+    let expectedToken = accessToken.id + ":" + sha1(accessToken.token + ":" + filename + ":" + timeStone);
```
Comment (dynamiccast) [inline api/policies/checkDownloadToken.js]: Isolate this in `StorageService`

File: api/services/PlazaService.js [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    let data = {
+      "command": cmd,
+      "stdin": stdin
+    };
+
+    let options = {
+      host: hostname,
+      path: '/exec',
+      port: '9090',
```
Comment (dynamiccast) [inline api/services/PlazaService.js]: Hard coded port

File: docker-compose-dev.yml [hunk 12]
```diff
@@ -54,3 +71,7 @@ services:
       - POSTGRES_USER=nanocloud
     network_mode: host
     container_name: "postgres"
+
+volumes:
```
Comment (dynamiccast) [inline docker-compose-dev.yml:77]: We should use file sytem mounted volumes in developement to be able to access any file any time.

File: tests/test-licences.sh [hunk 13]
```diff
@@ -121,3 +121,18 @@ for file in $(find tests -name "*.js"); do
     fi
   fi
 done
+
+echo "### Looking for source files in *plaza* directory"
+for file in $(find plaza -name "*.go"); do
+  echo "Scanning ${file}" >> add_licence.log
+  if [ -z "$(grep ${LICENCE_PATTERN} "${file}")" ]; then
+    if [ "tests" = "${ACTION}" ]; then
+      echo "Missing licence in : ${file}"
```
Comment (dynamiccast) [inline tests/test-licences.sh]: licence -> license

File: tests/unit/services/PlazaService.test.js [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+          .then((storage) => {
+            let file = {
+              filename: filename,
+              fd: "./tests/unit/services/" + filename
+            };
+            PlazaService.upload(
+                storage,
+                file,
+                () => {
+                  done();
```
Comment (dynamiccast) [inline tests/unit/services/PlazaService.test.js]: No expectation in this test ?

**Comment by dynamiccast** [comment]
Missing license in `plaza/vendor/github.com/gedex/inflector/example/example.go`.

File: api/services/StorageService.js [hunk 15]
```diff
@@ -57,5 +56,41 @@ module.exports = {
...
+    let timestamp = Date.now() / 1000;
+    let timeStone = timestamp + (3600 - timestamp % 3600);
+
+    return accessToken.id + ":" + sha1(accessToken.token + ":" + filename + ":" + timeStone);
+  },
+
+  /**
+   * checkToken
+   *
+   * Check wether a download token is valid or not
```
Comment (dynamiccast) [inline api/services/StorageService.js]: wether -> whether

File: api/policies/checkDownloadToken.js [hunk 16]
```diff
@@ -22,27 +22,23 @@
...
 
-  let timestamp = Date.now() / 1000;
-  let timeStone = timestamp + (3600 - timestamp % 3600);
-
-  AccessToken.findById(downloadToken.split(":")[0], (err, accessTokens) => {
+  AccessToken.findOne({
+    id: downloadToken.split(":")[0]
+  }, (err, accessToken) => {
     if (err !== null) {
       res.negotiate(err);
```
Comment (dynamiccast) [inline api/policies/checkDownloadToken.js]: `res.nogotiate` should be the last function. Please call with `return res.nogotiate`

File: api/policies/checkDownloadToken.js [hunk 17]
```diff
@@ -22,27 +22,23 @@
...
     }
-    let accessToken = accessTokens[0];
-    let expectedToken = accessToken.id + ":" + sha1(accessToken.token + ":" + filename + ":" + timeStone);
 
-    if (expectedToken !== downloadToken) {
+    if (StorageService.checkToken(accessToken, downloadToken, filename)) {
+      next();
```
Comment (dynamiccast) [inline api/policies/checkDownloadToken.js]: Please return at this point `return next()`

File: api/policies/checkDownloadToken.js [hunk 18]
```diff
@@ -22,27 +22,23 @@
...
+    } else {
       res.negociate(new Error("Wrong download token"));
```
Comment (dynamiccast) [inline api/policies/checkDownloadToken.js]: Should be the last function -> `return res.negotiate`

File: README.md (hunk 2 above)
Comment (Gentux) [inline README.md:30]: Done

File: api/policies/checkDownloadToken.js [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+module.exports = function(req, res, next) {
+  let filename = req.query["filename"];
+  let downloadToken = req.query["token"];
+
+  AccessToken.findOne({
+    id: downloadToken.split(":")[0]
+  }, (err, accessToken) => {
+    if (err !== null) {
+      res.negotiate(err);
```
Comment (dynamiccast) [inline api/policies/checkDownloadToken.js]: `res.negotiate` should be the last function ever call. Please precede every call with a `return` statement.

File: api/policies/checkDownloadToken.js [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    if (StorageService.checkToken(accessToken, downloadToken, filename)) {
+      return next();
+    } else {
+      return res.negociate(new Error("Wrong download token"));
```
Comment (dynamiccast) [inline api/policies/checkDownloadToken.js]: Wrong -> Invalid

File: tests/unit/services/StorageService.test.js [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+
+/* global AccessToken,ConfigService,StorageService,User */
+
+const expect = require('chai').expect;
+
+describe('Storage Service', function() {
+  before(function(done) {
+    ConfigService.set("storageAddress", "localhost")
+    .then(() => {
+      ConfigService.set("storagePort", 9090)
```
Comment (dynamiccast) [inline tests/unit/services/StorageService.test.js]: No need to break the chain of promise here.

File: docker-compose-dev.yml [hunk 22]
```diff
@@ -34,6 +34,23 @@ services:
     container_name: "proxy"
     restart: always
 
+  storage:
+    image: nanocloud/storage
+    network_mode: host
+    container_name: "storage"
```
Comment (dynamiccast) [inline docker-compose-dev.yml:40]: Needs `depends_on: plaza`

Comment (Gentux) [inline docker-compose-dev.yml:40]: done
